src/ZEO/asyncio/smp.py

```diff
@@ -0,0 +1,201 @@
+"""SizedMessageProtocol.
+
+The protocal exchanges ``bytes`` messages.
+On the wire, a message is represented by a 4 bytes length followed
+by the message's bytes.
+"""
+
+from .compat import asyncio
+from logging import getLogger
+import struct
+
+logger = getLogger(__name__)
+
+
+class SizedMessageProtocol(asyncio.Protocol):
+    """asyncio protocol for the exchange of sized messages.
+
+    A protocol object can be used as a connection.
+    Write methods are
+    ``write_message`` (write a single message)
+    and ``write_message_iter`` (write the messages generated by an iterator).
+    A received message is reported via the ``receive`` callback.
+
+    Messages are received in the same order as they have been
+    written.
+
+    ``SizedMessageProtocol`` instances can be used concurrently
+    from coroutines (executed in the same thread).
+    They are not thread safe.
+    """
+
+    transport = None
+
+    def __init__(self, receive):
+        self.receive = receive
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+        # output handling
+        pack = struct.pack
+        paused = []  # if ``paused`` we must buffer output
```

Comment:
Using a boolean would be more natural (than `[}`)

Reply:
`paused` indeed becomes boolean in `knext+kpy3` branch:

https://github.com/zopefoundation/ZEO/blob/e17cd2a43ee877f47f8d0c8ac2c3d564f7a32b5d/src/ZEO/asyncio/smp.py#L40-L42

But being boolean requires `nonlocal`

https://github.com/zopefoundation/ZEO/blob/e17cd2a43ee877f47f8d0c8ac2c3d564f7a32b5d/src/ZEO/asyncio/smp.py#L102-L104

which is available only on py3, not py2. So for compatibility with py2 this is list to which e.g. `pause_writing` appends 1.

Also: this is not my invention. `.paused` was list originally. And it appeared in `smp.py` as list by way of `SizedMessageProtocol` being refactored from `asyncio.base` in e89c4bc4a3e2683d09f269970690e9f6b4ad68bd.